text/0000-type-inference.md

```diff
@@ -0,0 +1,168 @@
+type injected into `cleanup` and `middleware`. All of this can be done without having any return type from `createPlugin`.
+
+The second part is type checking the `app.register` and `app.enhance` calls. This is done to verify the type of the token matches the provides type of the plugin, 
+and this is the part which depends on the return type of `createPlugin`. The key thing to notice here is that we only need the type of the `provides` method of 
+the plugin to verify it matches the token. At this point, deps, cleanup, and middleware are all irrelevant. 
+
+Because `createPlugin` should return an opaque type and because we only need the `provides` type to verify with `app.register`, we can update the `createPlugin` 
+return type to remove all generics in input positions without losing any type safety for consumers. There are several ways we could do this, but the simplest
+would be replacing the `service` generic with `any` in the return type of `createPlugin` whenever it is referenced in an input position. We also need to specify
+the properties as covariant.
+
+```js
+type CreatePluginArgs<+Deps, +Service> = {|
+  +deps?: Deps,
+  +provides?: (Deps: $ObjMap<Deps & {}, ExtractReturnType>) => Service,
+  +middleware?: (
+    Deps: $ObjMap<Deps & {}, ExtractReturnType>,
+    Service: Service
+  ) => Middleware,
+  +cleanup?: (service: Service) => Promise<void>,
+|};
+
+type FusionPlugin<+Deps, +Service> = {|
+  +deps?: Deps,
+  +provides?: (Deps: $ObjMap<Deps & {}, ExtractReturnType>) => Service,
+  +middleware?: (
+    Deps: $ObjMap<Deps & {}, ExtractReturnType>,
+    Service: any 
+  ) => Middleware,
+  +cleanup?: (service: any) => Promise<void>,
+|};
+
+function createPlugin<Deps, Service>(p: CreatePluginArgs<Deps, Service>): FusionPlugin<Deps, Service> {
+  // ...
+}
+```
+
+This will allow us to get type inference with fusion plugins across module boundaries while still getting all the performance benefits of the types first
+flow architecture and keeping all the same type safety for fusion consumers.
+
+# Drawbacks
+
+While there is type information lost, I would argue it is not type information that should be used in any meaningful way as it is not recommended
+to directly access properties of fusion plugins. 
+
+The only other drawback we could potentially see is if future changes to flow or the fusion plugin API cause us to lose the ability to get type inference
+we will need to migrate all usages of `createPlugin` to include explicit type annotations, as we have done in the past.
+
+---
+
+# Alternatives
+
+There are various alternatives to how we can hack the type system to remove references to generics in input positions. One possibility is we could completely remove
+the presence of the `cleanup` and `middleware` properties from the return type of `createPlugin`. I think this is worse than using `any` since those properties 
```

Comment:
Agreed!  This would also require some hackery in `fusion-core` if we went this route.